main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala

```diff
@@ -49,6 +50,17 @@ object CodeHinter {
     deprecated ++ experimental
   }
 
+  /**
+    * Computes code quality hints for the given class `typeclass`.
+    */
+  private def visitClass(typeclass: TypedAst.Class)(implicit root: Root): List[CodeHint] = {
```

Comment:
This sneaked by me in the last review, but your image made me realize it.

We don't actually want to display a hint where a class is declared, but rather where it is _used_. For example, 
```scala

@Deprecated
lawless class OldClass[a] // no hint here

pub def doThing(): a with OldClass[a]
                          ^^^^^^^^
                          deprecated
```

Same thing for enums.

Reply:
Oh I see, you are right!

Reply:
I am facing a problem in doing so, mainly:

I should be able to extract the enum from an expression. For example in `Def` we currently use `Type`s and I find difficult to get back to the original enum info. 

Also I think this is kind of a broader issue that makes the vscode extension not precise when hovering on an element:

![Screenshot from 2022-03-16 22-36-01](https://user-images.githubusercontent.com/55001950/158695196-ebb21701-a0cf-489d-b91b-50aa7eef94a1.png)

Do you have any suggestions?

Reply:
I'm not sure I totally understand your question, but I'll explain some things and hopefully point you in the right direction :)

We want to highlight `Tree` in your example `get` function. In the `CodeHinter`, it looks like `visitDef` only looks at the expression, so we'll need to add some processing of the formal parameters (`defn.spec.fparams`). For each of these, we can explore the type. If we ever hit an `Enum` type, then we can grab its symbol and look it up in the root (`root.enums`). Then we can look at that enum's annotations, and see if it's deprecated, etc.

Does that answer your question?

Reply:
So by looking better at the issue, it seems like the `CodeHinter` should have access also to the information given in `lsp/Index` because here we want the hint to be displayed at usage site.

I am also wondering why not to put this phase as a provider in LSP as it is concerned with computing information for the editor which is not displayed at compilation time only (the only place where we call `run` is indeed the LanguageServer).

Reply:
@magnus-madsen  I am a bit confused, is `flix` argument of the CodeHinter.run method used actually?

Because otherwise I would remove it and just refactor this file as a provider with a similar run method as the other ones have.

Reply:
> So by looking better at the issue, it seems like the `CodeHinter` should have access also to the information given in `lsp/Index` because here we want the hint to be displayed at usage site.

I don't think that's necessary in this case. Take a look at what `visitExp` does to handle annotated defs.

> I am also wondering why not to put this phase as a provider in LSP as it is concerned with computing information for the editor which is not displayed at compilation time only (the only place where we call `run` is indeed the LanguageServer).

I'm not 100% sure, but I believe at least part of the reason is that we've considered in the past making this info available from the command line without launching the server.

Reply:
Thanks for the suggestions! I will then take a closer look :)

Reply:
So, my problem is how to get the enum symbol from the formal parameters :sweat_smile:

Reply:
Ah, I forgot about TypeConstructor!

Once you have a Type, you can match on it. You'll need to search through the whole type tree. (In particular, looking at both arguments to `Type.Apply`. What you're looking for is `Type.Cst(TypeConstructor.KindedEnum(sym, _), _)`. Then you can look up that `sym` in `root.enums` and find if it's annotated.

Reply:
Cool thanks a lot for the explanation :)

Reply:
I have submitted another PR addressing the issue for enums first to see if the approach is ok, then I will resume to work on this one :)

Reply:
I am merging this PR now, and then I will add some other comments to the other PR.